profile_xf05id1/startup/11-secret_undulator.py

```diff
@@ -0,0 +1,387 @@
+    'Upstream upper motor on SRX undulator'
+    readback = Cpt(EpicsSignal, '}REAL_POSITION_US_UPPER')
+    setpoint = Cpt(EpicsSignal, '-Mtr:6}Inp:Pos')
+    actuate = Cpt(EpicsSignal, '-Mtr:6}Sw:Go')
+    done = Cpt(UVDoneMOVN, None, moving='moving',
+               readback='readback', add_prefix=())
+    stop_signal = Cpt(EpicsSignal, '-Mtr:6}Pos.STOP')
+
+    moving = Cpt(EpicsSignal, '-Mtr:3}Pos.MOVN')
+
+
+class UndlatorMotorUSL(UndulatorPositioner):
+    'Upstream lower motor on SRX undulator'
+    readback = Cpt(EpicsSignal, '}REAL_POSITION_US_LOWER')
+    setpoint = Cpt(EpicsSignal, '-Mtr:8}Inp:Pos')
+    actuate = Cpt(EpicsSignal, '-Mtr:8}Sw:Go')
+    done = Cpt(UVDoneMOVN, None, moving='moving',
+               readback='readback', add_prefix=())
+    stop_signal = Cpt(EpicsSignal, '-Mtr:8}Pos.STOP')
+
+    moving = Cpt(EpicsSignal, '-Mtr:8}Pos.MOVN')
+
+
+class UndlatorMotorDSU(UndulatorPositioner):
+    'Downstream upper motor on SRX undulator'
+    readback = Cpt(EpicsSignal, '}REAL_POSITION_DS_UPPER')
+    setpoint = Cpt(EpicsSignal, '-Mtr:5}Inp:Pos')
+    actuate = Cpt(EpicsSignal, '-Mtr:5}Sw:Go')
+    done = Cpt(UVDoneMOVN, None, moving='moving',
+               readback='readback', add_prefix=())
+    stop_signal = Cpt(EpicsSignal, '-Mtr:5}Pos.STOP')
+
+    moving = Cpt(EpicsSignal, '-Mtr:5}Pos.MOVN')
+
+
+class UndlatorMotorDSL(UndulatorPositioner):
+    'Downstream lower motor on SRX undulator'
+    readback = Cpt(EpicsSignal, '}REAL_POSITION_DS_LOWER')
+    setpoint = Cpt(EpicsSignal, '-Mtr:7}Inp:Pos')
+    actuate = Cpt(EpicsSignal, '-Mtr:7}Sw:Go')
+    done = Cpt(UVDoneMOVN, None, moving='moving',
+               readback='readback', add_prefix=())
+    stop_signal = Cpt(EpicsSignal, '-Mtr:7}Pos.STOP')
+
+    moving = Cpt(EpicsSignal, '-Mtr:4}Pos.MOVN')
+
+
+class PowerUndulator(Device):
+    'Simple aggregate device to hold undulator motors'
+    us_lower = Cpt(UndlatorMotorUSL, '')
+    us_upper = Cpt(UndlatorMotorUSU, '')
+    ds_lower = Cpt(UndlatorMotorDSL, '')
+    ds_upper = Cpt(UndlatorMotorDSU, '')
+
+pu = PowerUndulator('SR:C5-ID:G1{IVU21:1', name='pu')
+
+
+class UTemperatures(Device):
+    'Undulator temperatures'
+    T1 = Cpt(EpicsSignal, '-Pt:1}T')
+    T2 = Cpt(EpicsSignal, '-Pt:2}T')
+    T3 = Cpt(EpicsSignal, '-Pt:3}T')
+    T4 = Cpt(EpicsSignal, '-Pt:4}T')
+    T5 = Cpt(EpicsSignal, '-Pt:5}T')
+    T6 = Cpt(EpicsSignal, '-Pt:6}T')
+    T7 = Cpt(EpicsSignal, '-Pt:7}T')
+    T8 = Cpt(EpicsSignal, '-Pt:8}T')
+    T9 = Cpt(EpicsSignal, '-Pt:9}T')
+    T10 = Cpt(EpicsSignal, '-Pt:10}T')
+    T11 = Cpt(EpicsSignal, '-Pt:11}T')
+    T12 = Cpt(EpicsSignal, '-Pt:12}T')
+    T13 = Cpt(EpicsSignal, '-Pt:13}T')
+    T14 = Cpt(EpicsSignal, '-Pt:14}T')
+    T15 = Cpt(EpicsSignal, '-Pt:15}T')
+    T16 = Cpt(EpicsSignal, '-Pt:16}T')
+
+ut = UTemperatures('SR:C5-ID:G1{IVU21:1')
+
+
+
+
+@wrap_with_decorator(run_wrapper)
+def ud_crab_plan(pu, us_u, us_l, ds_u, ds_l, other_dets=None):
```

Comment:
If the upper or lower motor pairs get more that .08 - .1 different the undulator will lock out control at the controller level (to prevent the hardware from ripping it's self apart).  Moving the undulator can dump the beam (by bumping it outside of the safety envelope by changing the magnetic field which causes the machine to dump it's self) or screw with other users by moving their beam.